### PR TITLE
Project 1

### PR DESCRIPTION
**Pull request by luisesanmartin**
This is my first commit of project 1. I'm still pending to implement test functions for each task, I plan to have it finalized during the week. These are my comments on how the project is now:
- I think it's too long to be finalized in 2.5 hours. That's not necessarily bad, as participants can try to finish as much as they can during the session and then complete it as an assignment if they want. My only concern with that is that it might affect motivation
- In that sense, a few changes to make it easier could be to provide the `neighborhood()` function ourselves (I think that's going to be one of the bottlenecks) and/or change the moving rules to something simpler, like agents moving to any available location at random

### CODE REVIEW

**Comment by luisesanmartin** [comment]
@kbjarkefur I also just realized that a few more steps are needed for importing and running files committed to GH with a colab notebook (instructions [here](https://stackoverflow.com/questions/60664979/how-to-use-google-colab-to-run-py-files-from-github-repositories)). I'm going to add this so that participants can import all files to the notebook

**Comment by weilu** [comment]
It's a fun coding task. Took me 3 hours to finish as-is. I added some basic test cases to each step to check correctness. Not comprehensive so [my solution](https://github.com/worldbank/dec-python-course/blob/project1-wei/1-foundations/project-1-schellings-model/Project%201%20-%20Schelling's%20model%20of%20segregation.ipynb) is not guaranteed to be fully correct. 

The only liberty I took was with the return value of `neighborhood` – instead of the sliced array of arrays, I just returned the top left and bottom right locations marking the neighborhood. I can't explain exactly why I just intuitively did that; perhaps just the light-weight-ness and the consistency of working with indexes?

Here are some notes I took while going through the project:

- reading/importing relative files does not work in colab. Need to fix
- Why are types represented using 1 and 2 and not 0 and 1? I feel like it's a missed opportunity to help people get used to the binary world ;) Perhaps the parser can also convert them to integers so they are slightly easier to work with (e.g. not having to type quotes for test cases).
- var name s/location1/current_location s/location2/new_location for readability
- "the number of rounds that passed in the simulation" ambiguous. Does it include the no-move round?

**Comment by weilu** [comment]
**Option to simplify the task**: drop the `extension` parameter and fix it at 1
**Option to add bonus/advanced material**: ask the students to think of ways to optimize program: are there things that can be done faster? E.g. separating the candidate listing and moving distance calculation is a good conceptual model for the first pass but the search can be optimized by starting from the cells closer to the agent then expanding outward (increasing moving distance) and terminating the search for more candidates as soon as the nearest above-threshold location is found.

I personally find the square matrix assumption somewhat arbitrary. It's an assumption that doesn't simplify things much. Maybe there are theory reasons behind it, but I don't see any from a programming point of view.

File: 1-foundations/project-1-schelling-model/Project 1 - Schelling's model of segregation.ipynb [hunk 1]
```diff
@@ -0,0 +1,676 @@
...
+    "## Part 2: Python implementation\n",
+    "\n",
+    "This is how our specific implementation of Schelling's model will work:\n",
+    "\n",
+    "1. The initial state of the city (the grid) will be read from pre-defined files\n",
+    "1. When a round of the simulation begins, start the iteration with the square at the top left and perform the following operations:\n",
+    "    - Evaluate if the agent is dissatisfied\n",
+    "    - If they are:\n",
+    "        - Move the agent to its corresponding lot applying the moving rule\n",
+    "        - Update the city\n",
```
Comment (kbjarkefur) [inline 1-foundations/project-1-schelling-model/Project 1 - Schelling's model of segregation.ipynb:111]:
```suggestion
    "1. When a round of the simulation begins, start the iteration with the square at the top left and perform the following operations:\n",
    "    - Evaluate if there is an agent in the square\n",
    "    - If there is one:\n",
    "        - Evaluate if the agent is dissatisfied\n",
    "        - If they are:\n",
    "            - Move the agent to its corresponding lot applying the moving rule\n",
    "            - Update the city\n",
```

File: 1-foundations/project-1-schelling-model/Project 1 - Schelling's model of segregation.ipynb [hunk 2]
```diff
@@ -0,0 +1,676 @@
...
+    "    - If no one moved in the round, finish the simulation\n",
+    "    - If more than T rounds have passed, finish the simulation\n",
+    "1. If the simulation didn't finish, continue with a new round of the model\n",
+    "\n",
+    "And this is the moving rule we'll use:\n",
+    "\n",
+    "- An agent will move to another location that has a higher or equal similarity than the acceptable threshold. If no location has a higher or equal similarity, they will not move\n",
+    "- If more than one location is acceptable, they will move to the closest one in terms of Euclidean distance (formula is below)\n",
+    "- If more than one acceptable location is the closest, they will move to the one that has the highest similarity\n",
+    "- If still more than one is a candidate location, they will be indifferent between them and will choose any\n",
```
Comment (kbjarkefur) [inline 1-foundations/project-1-schelling-model/Project 1 - Schelling's model of segregation.ipynb:124]: What if the agent is dissatisfied but there are no locations where the acceptable threshold is met?

File: 1-foundations/project-1-schelling-model/Project 1 - Schelling's model of segregation.ipynb [hunk 3]
```diff
@@ -0,0 +1,676 @@
...
+    "1. When the rows finishes, continue with the next row\n",
+    "1. When all the rows are looped over, perform the following checks:\n",
+    "    - If no one moved in the round, finish the simulation\n",
+    "    - If more than T rounds have passed, finish the simulation\n",
+    "1. If the simulation didn't finish, continue with a new round of the model\n",
+    "\n",
+    "And this is the moving rule we'll use:\n",
+    "\n",
+    "- An agent will move to another location that has a higher or equal similarity than the acceptable threshold. If no location has a higher or equal similarity, they will not move\n",
+    "- If more than one location is acceptable, they will move to the closest one in terms of Euclidean distance (formula is below)\n",
```
Comment (kbjarkefur) [inline 1-foundations/project-1-schelling-model/Project 1 - Schelling's model of segregation.ipynb:122]:
```suggestion
    "- An agent will move to another location that has a higher or equal similarity than the acceptable threshold. If no location has a higher or equal similarity, they will not move\n",
    "- An agent may only move to locations currently not occupied by any other agents.\n",
    "- If more than one location is acceptable, they will move to the closest one in terms of Euclidean distance (formula is below)\n",
```

**Comment by kbjarkefur** [comment]
can we change the file name to `project-1-schelling-model-of-segregation.ipynb`? Otherwise the URL gets so ugly. https://github.com/worldbank/dec-python-course/blob/project1/1-foundations/project-1-schelling-model/Project%201%20-%20Schelling's%20model%20of%20segregation.ipynb

File: 1-foundations/project-1-schelling-model/Project 1 - Schelling's model of segregation.ipynb [hunk 4]
```diff
@@ -0,0 +1,676 @@
...
+    "import testing"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Project 1: An implementation of Schelling's model of segregation in Python\n",
+    "\n",
+    "The first project of the course consists of a simplified implementation of a classic example of agent-based modeling: Schelling's model of segreagation. You will implement a simplified version of the model in Python using only custom functions and basic data types. The project is divided in tasks that should be solved in order and build upon each other until a final function that runs an entire simulation of the model. To help making sure that each task is solved correctly before moving on to the next one, we have prepared testing functions that your solution must pass before a task is completed.\n",
```
[the remark this answers is not in the document]
Comment (luisesanmartin) [reply 1-foundations/project-1-schelling-model/Project 1 - Schelling's model of segregation.ipynb]: added in b9d6c24

File: 1-foundations/project-1-schelling-model/Project 1 - Schelling's model of segregation.ipynb (hunk 1 above)
Comment (luisesanmartin) [reply 1-foundations/project-1-schelling-model/Project 1 - Schelling's model of segregation.ipynb]: I pushed other changes before committing the suggestions, but recreated them and pushed in b9d6c24

File: 1-foundations/project-1-schelling-model/Project 1 - Schelling's model of segregation.ipynb (hunk 3 above)
Comment (luisesanmartin) [reply 1-foundations/project-1-schelling-model/Project 1 - Schelling's model of segregation.ipynb]: applied in b9d6c24

**Pull request merged**